### PR TITLE
feat: create bridge configuration in offline mode

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
This PR adds a `--offline` flag to `tedge connect c8y` that skips connection checks and device registration, which allows the creation of bridge configuration and enables all services while offline.

TODO:

- [x] update doc about new command
- [x] add system test

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
* #2831 
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by gligorisaev — COMMENTED** [review]
Documentation:

- The command syntax is clear and concise. It provides the necessary information for the user to understand how to execute the command.

Test:

- The structure of the test case is clear and consistent, with each step clearly outlined.
- Test case names are descriptive and convey the intended functionality.

File: docs/src/start/connect-c8y.md [hunk 1]
```diff
@@ -248,6 +248,54 @@ Persisting tedge-agent on reboot.
...
+Checking if tedge-agent is installed.
+
+Starting tedge-agent service.
+
+Persisting tedge-agent on reboot.
+
+tedge-agent service successfully started and enabled!
+```
+
+In order for the above command to work, the device must be pre-registered in your tenant. If the internet connection is restored, the thin-edge.io instance will establish a connection with the cloud.
```
Comment (didier-wenzek) [inline docs/src/start/connect-c8y.md:297]: As there is no attempt to connect the cloud, `tedge connect c8y --offline` should work even if the device is not registered yet.

```suggestion
If the device certificate is trusted by c8y when the internet connection is restored,  the thin-edge.io instance will  automatically establish a connection with the cloud.
```

File: tests/RobotFramework/tests/tedge_connect/tedge_connect_offline.robot [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+Test Tags           theme:mqtt    theme:c8y    adapter:docker
+
+
+*** Test Cases ***
+Check offline bootstrapping
+    ThinEdgeIO.Execute Command    sudo tedge disconnect c8y
+    ThinEdgeIO.Service Should Be Stopped    tedge-mapper-c8y
+
+    ThinEdgeIO.Disconnect From Network
```
Comment (didier-wenzek) [inline tests/RobotFramework/tests/tedge_connect/tedge_connect_offline.robot:16]: Where is defined this test command?

**Review by didier-wenzek — COMMENTED** [review]
This correctly implements what is described by https://github.com/thin-edge/thin-edge.io/issues/2831.

However, looking at the code and the command help, I find the `connect` command not more complicated than required. One cause of this complexity is that the bridge file is removed when the connection check fails.

What about:
- stop removing the bridge configuration on failed connection check
- perform no connection check on connect unless explicitly requested with `--test`.

So:
- `tedge connect c8y` would behave as `tedge connect c8y --offline` as implemented here
- `tedge connect c8y --test` would behave as `tedge connect c8y` as currently implemented when there is no c8y bridge configuration and as `tedge connect c8y --test` when there is a c8y bridge configuration.

File: tests/RobotFramework/tests/tedge_connect/tedge_connect_offline.robot (hunk 2 above)
Comment (gligorisaev) [reply tests/RobotFramework/tests/tedge_connect/tedge_connect_offline.robot:16]: It is part of the DeviceLibrary.py

Comment (reubenmiller) [reply tests/RobotFramework/tests/tedge_connect/tedge_connect_offline.robot:16]: Under the hood the "disconnect" command is provided by the test adapter. For the container adapter, it uses docker api to disconnect the container from the docker network, https://docker-py.readthedocs.io/en/stable/networks.html#docker.models.networks.Network.disconnect

The container adapter code is reference [here](https://github.com/reubenmiller/device-test-core/blob/29af60f38e04427dc1cc5816229f375619de7f30/device_test_core/docker/device.py#L223)

Comment (didier-wenzek) [reply tests/RobotFramework/tests/tedge_connect/tedge_connect_offline.robot:16]: Thanks - This is clearer

File: crates/core/tedge/src/cli/connect/cli.rs [hunk 3]
```diff
@@ -39,25 +43,31 @@ pub enum TEdgeConnectOpt {
...
                 is_test_connection,
+                offline_mode,
                 service_manager: service_manager(&context.config_location.tedge_config_root_path)?,
             },
             TEdgeConnectOpt::Az { is_test_connection } => ConnectCommand {
                 config_location: context.config_location.clone(),
                 config: context.load_config()?,
                 cloud: Cloud::Azure,
                 is_test_connection,
+                offline_mode: false,
```
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/connect/cli.rs:62]: Why no `offline_mode` for Az and AWS?

File: docs/src/start/connect-c8y.md (hunk 1 above)
Comment (Ruadhri17) [reply docs/src/start/connect-c8y.md]: resolved in ad228b8049743cff3f0b8a463392554dceb7fd8c

File: crates/core/tedge/src/cli/connect/cli.rs (hunk 3 above)
Comment (reubenmiller) [reply crates/core/tedge/src/cli/connect/cli.rs]: We don't have to worry about supporting `--offline` for the `tedge reconnect <mapper>` as the command does not support.

If users want to regenerate the bridge configuration, then they can already do so using `tedge refresh-bridges` which will rebuild the existing bridge files without having to test the cloud connection.

**Review by reubenmiller — APPROVED** [review]
Approved. I pushed some minor doc changes (rewording etc.) in 746cbe249b671b918d9691965be10ae420724ce8

**Pull request merged**